src/context/typecore.ml

```diff
@@ -222,7 +222,31 @@ let add_local ctx k n t p =
 	ctx.locals <- PMap.add n v ctx.locals;
 	v
 
+let check_identifier_name ctx name kind p =
+	if starts_with name '$' then
+		display_error ctx (kind ^ " names starting with a dollar are not allowed") p
+	else
+		let is_valid_identifier s =
+			match s with
+			| "new" | "this" -> true
```

Comment:
This probably should take into account all keywords. Not only `new` and `this`

Reply:
Or just extract the keyword-independent part of `Lexer.is_valid_identifier` into a separate function

Reply:
The thing is, we _don't want to allow other keywords as identifiers. These are just two special cases that we have to allow.

Reply:
`new` is special for field declaration and hence only should be allowed for `kind = "field"`. Why do we want `this` to be allowed?
Also, if I make `var new;` will it pass the check?

Reply:
Good point about `new`.

There were errors all over the place when not allowing `this`... I suppose there must be actual variables called that and then escaped to `this later?

Reply:
Probably `this` should be allowed for a specific `tvar_origin` case.

Reply:
or for `VGenerated` from `tvar_kind`

Reply:
Doesn't this only check `VUser` vars?

`this` is used for abstracts.

Reply:
> Doesn't this only check VUser vars?

If so, why did it "error all over the place"?

Reply:
Hm... looks like `add_local_with_origin` is indeed being called with `n = "this"`, even though that implies `VUser`.

Reply:
The call in `TypeloadFunction.type_function` is an example of that.